File: leetcode.go
package main

import (
	"io/ioutil"
	"os"
	"path"
	"fmt"
	"strings"
)

type Solution struct {
	DirName string
	DirUrl  string
	Urls    map[string]string
}

const LEETCODE_DIR = `./LeetCode`

var suffix = []string{"cpp", "java", "js", "python", "go", "kt"}

func main() {
	md, err := os.OpenFile("LeetCode.md", os.O_CREATE|os.O_WRONLY, 0666)
	defer md.Close()
	panicErr(err)

	list, err := ioutil.ReadDir(LEETCODE_DIR)
	panicErr(err)
	solutions := make([]Solution, len(list))
	for idx, dir := range list {
		if dir.Name() == ".vscode" {
			continue
		}
		if dir.IsDir() {
			solution := Solution{DirName: dir.Name(), Urls: make(map[string]string)}
			solution.DirUrl = LEETCODE_DIR + "/" + dir.Name()
			files, err := ioutil.ReadDir(solution.DirUrl)
			panicErr(err)
			for _, f := range files {
				if dir.Name() == ".vscode" {
					continue
				}
				ext := path.Ext(f.Name())[1:]
				if ok := contains(suffix, ext); ok {
					solution.Urls[ext] = LEETCODE_DIR + "/" + dir.Name() + "/" + f.Name()
				}
			}
			solutions[idx] = solution
		}
	}

	md.WriteString("Problem | Cpp | Java | JS | Python | Go | Kotlin\n")
	md.WriteString("---|---|---|---|---|---|---\n")
	for _, s := range solutions {
		md.WriteString(fmt.Sprintf("[%s](%s)", s.DirName, strings.Replace(s.DirUrl, " ", "%20", -1)))
		for _, suf := range suffix {
			md.WriteString("|")
			if _, ok := s.Urls[suf]; ok {
				md.WriteString(fmt.Sprintf("[%s](%s)", suf, strings.Replace(s.Urls[suf], " ", "%20", -1)))
			}
		}
		md.WriteString("\n")
	}
}

func contains(source []string, target string) bool {
	for _, t := range source {
		if t == target {
			return true
		}
	}
	return false
}

func panicErr(err error) {
	if err != nil {
		panic(err)
	}
}

File: LeetCode/leetcode.py
import os
import re

solution = {
    'dir_name' : '',
    'dir_url'  : '',
    'file_urls': {}
}

suffix = ["cpp", "java", "js", "py", "go", "kt"]
leetcode_loc = '.'

def main():
    solution_list = []
    for dir_url, _, files in os.walk(leetcode_loc):
        dir_name = dir_url.split('\\')[-1]
        if re.match(r'^[0-9]{4}\. .*', dir_name):
            solution['dir_name'] = dir_name
            solution['dir_url']  = os.path.join('LeetCode', dir_url[2:]) 
            solution['file_urls'] = {f.split('.')[-1] : os.path.join(dir_url, f) for f in files}
            solution_list.append(solution.copy())

    with open('./LeetCode.md', 'w+', encoding="utf-8") as md:
        md.write("# Leetcode\n**Problems have been solved:** " + str(len(solution_list)))
        md.write("\n\n")
        md.write("Problem | Cpp | Java | JS | Python | Go | Kotlin\n")
        md.write("---|---|---|---|---|---|---\n")

        for s in solution_list:
            md.write("[{}]({})".format(s['dir_name'], s['dir_url'].replace(" ", "%20")).replace("\\", "/"))
            for ext in suffix:
                md.write('|')
                if ext in s['file_urls']: 
                    md.write("[{}]({})".format(ext, s['file_urls'][ext].replace(" ", "%20").replace("\\", "/")))
            md.write('\n')
        md.write("\n\n------\n\n")

if __name__ == '__main__':
    main()
    